workflow/controller/taskset.go

```diff
@@ -0,0 +1,204 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/pointer"
+
+	"github.com/argoproj/argo-workflows/v3/errors"
+	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	errorsutil "github.com/argoproj/argo-workflows/v3/util/errors"
+	"github.com/argoproj/argo-workflows/v3/util/retry"
+	waitutil "github.com/argoproj/argo-workflows/v3/util/wait"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+)
+
+func (woc *wfOperationCtx) getAgentPodName() string {
+	return woc.wf.NodeID("agent") + "-agent"
+}
+
+func (woc *wfOperationCtx) executeTaskSet(ctx context.Context, nodeName string, templateScope string, tmpl *wfv1.Template, orgTmpl wfv1.TemplateReferenceHolder, opts *executeTemplateOpts) (*wfv1.NodeStatus, error) {
+	node := woc.wf.GetNodeByName(nodeName)
+	if node == nil {
+		node = woc.initializeExecutableNode(nodeName, wfv1.NodeTypeHTTP, templateScope, tmpl, orgTmpl, opts.boundaryID, wfv1.NodePending)
+	}
+	mainCtr := woc.newExecContainer(common.MainContainerName, tmpl)
+	mainCtr.Command = []string{"argoexec", "agent"}
+	// Append Agent in end of pod name to differentiate from normal podname
+	podName := woc.getAgentPodName()
+	_, err := woc.createAgentPod(ctx, podName, []apiv1.Container{*mainCtr}, tmpl)
+	if err != nil {
+		return woc.requeueIfTransientErr(err, node.Name)
+	}
+	if node.Phase == wfv1.NodePending {
+		err := woc.controller.taskSetManager.CreateTaskSet(ctx, woc.wf, node.ID, *tmpl)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return node, nil
+}
+
+func (woc *wfOperationCtx) DeleteAgentPod(ctx context.Context) error {
+	agentPodName := woc.getAgentPodName()
+	log.WithField("workflow", woc.wf.Name).WithField("namespace", woc.wf.Namespace).Infof("Deleting Agent Pod")
+	var err error
+	err = waitutil.Backoff(retry.DefaultRetry, func() (bool, error) {
+		err = woc.controller.kubeclientset.CoreV1().Pods(woc.wf.Namespace).Delete(ctx, agentPodName, metav1.DeleteOptions{})
+		return apierr.IsNotFound(err) || err == nil, err
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (woc *wfOperationCtx) taskSetReconciliation() error {
+	taskSet, err := woc.getWorkflowTaskSet()
+	if err != nil {
+		return err
+	}
+	if taskSet == nil || taskSet.Status.Nodes == nil || len(taskSet.Status.Nodes) == 0 {
```

Comment:
if taskSet.Status.Nodes is nil, then `len(taskSet.Status.Nodes)  == 0` so you don't need the second or condition here - is there a test for this?